File: CreationalPatterns/AbstractFactory/abstract_products.py
from abc import ABC, abstractmethod


class ToyAnimal(ABC):
    @abstractmethod
    def interact(self):
        pass


class AquaticAnimal(ToyAnimal):
    pass


class TerrestrialAnimal(ToyAnimal):
    pass


class AvianAnimal(ToyAnimal):
    pass

File: CreationalPatterns/AbstractFactory/abstract_products.cpp
#include <iostream>

class ToyAnimal {
public:
    virtual ~ToyAnimal() {} // Virtual destructor
    virtual void interact() = 0; // Pure virtual function
};

class AquaticAnimal : public ToyAnimal {
public:
    void interact() override {
        std::cout << "Aquatic animal interaction" << std::endl;
    }
};

class TerrestrialAnimal : public ToyAnimal {
public:
    void interact() override {
        std::cout << "Terrestrial animal interaction" << std::endl;
    }
};

class AvianAnimal : public ToyAnimal {
public:
    void interact() override {
        std::cout << "Avian animal interaction" << std::endl;
    }
};
